// react-native-orion/lib/component/button/pushbutton.ios.js
/**
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

import React, { Component } from 'react';
import {
    StyleSheet,
    Text,
    View,
    ColorPropType,
    TouchableOpacity,
} from 'react-native';

import {
    Aspect,
    OrionText,
    OrionComponent,
} from 'react-native-lark';

const styles = StyleSheet.create({
    title:{
        textAlign: 'center',
        color: '#000000',
        width: Aspect.ratio(40),
        height: Aspect.ratio(20),
        borderColor: "#000000",
        textAlign: 'center',
        borderWidth: 1,
        fontSize: Aspect.ratio(14),
        lineHeight: Aspect.ratio(14),
        paddingVertical:  Aspect.ratio(3),
        backgroundColor: "transparent",
        overflow: 'hidden',
    },
});

const OrionPushButton = React.createClass({
    propTypes: {
        ...OrionText.propTypes,
        l_height: React.PropTypes.number,
        onPress: React.PropTypes.func,
    },
    getDefaultProps: function() {
        return {
            ...OrionText.defaultProps,
            l_height: Aspect.ratio(20),
        };
    },

    render:function (){
        let { l_height, l_text, style, ...props } = this.props;

        var fs = style.fontSize ? style.fontSize : Aspect.ratio(14);
        var textProp ={};
        textProp.height = l_height;
        textProp.fontSize = fs;
        textProp.lineHeight = fs;
        textProp.paddingVertical = (l_height - fs )/ 2;

        return (
            <TouchableOpacity
                activeOpacity={1.0}
                onPress={this.props.onPress}>
                <OrionText {...props} l_text={l_text} style={[styles.title, style, textProp]}/>
            </TouchableOpacity>
        );
    }
});


module.exports = OrionPushButton;
