// react-native-orion/lib/component/carousel/carouselPropType.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
'use strict';

import React from 'react';
import {
    View,
    ColorPropType
} from 'react-native';

import OrionComponent from  '../component';

const OrionCarouselPropType = React.PropTypes.shape({
    ...OrionComponent.propTypes,
    style: View.propTypes.style,
    children: React.PropTypes.node.isRequired,
    l_delay: React.PropTypes.number,
    l_pageStyle: View.propTypes.style,
    l_contentContainerStyle: View.propTypes.style,
    l_autoplay: React.PropTypes.bool,
    l_width: React.PropTypes.number,
    l_height: React.PropTypes.number,
    l_initPage:  React.PropTypes.number,
    l_loop: React.PropTypes.bool,
    l_pageControl: React.PropTypes.shape({
        enable: React.PropTypes.bool,
        color: ColorPropType,
        highlightColor: ColorPropType,
    }),
});

module.exports = OrionCarouselPropType;
