// react-native-orion/lib/component/listview/refreshlistviewPropType.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
'use strict';

import React from 'react';
import OrionComponent from '../component';

const OrionRefreshListViewPropType = React.PropTypes.shape({
    ...OrionComponent.propTypes,
    l_footerHeight : React.PropTypes.number,
    l_pullDistance : React.PropTypes.number,
    l_renderEmptyRow : React.PropTypes.func,
    l_renderHeaderRefreshIdle : React.PropTypes.func,
    l_renderHeaderWillRefresh : React.PropTypes.func,
    l_renderHeaderRefreshing : React.PropTypes.func,
    l_renderFooterInifiteIdle : React.PropTypes.func,
    l_renderFooterWillInifite : React.PropTypes.func,
    l_renderFooterInifiting : React.PropTypes.func,
    l_renderFooterInifiteLoadedAll : React.PropTypes.func,
});

module.exports = OrionRefreshListViewPropType;
