// react-native-orion/lib/component/navigator/navigatorimagebtn.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    StyleSheet,
    View,
    Image,
} from 'react-native';

import OrionComponent from '../component';
import OrionImage from '../image/image';
import OrionText from '../text/text';

const OrionNavigatorImageButton = React.createClass({

    propTypes: {
        ...OrionComponent.propTypes,
        style: View.propTypes.style,
        imageStyle: OrionImage.propTypes.style,
        imageSoure: Image.propTypes.source,
        titleStyle: OrionText.propTypes.style,
        title: React.PropTypes.string,
        layout: React.PropTypes.oneOf(['left','right']),
    },
    getDefaultProps: function() {
        return {
            ...OrionComponent.defaultProps,
            layout: 'left',
        };
    },
    render: function () {
        var {layout, imageStyle, imageSoure, titleStyle, title, style,...props} = this.props;
        var content;

        var icon_n = imageSoure ? (<Image  style={[styles.icon, imageStyle]} source={imageSoure} />) : null;
        var title_n = title ? (<OrionText  style={[styles.title, titleStyle]} l_text={title} />) : null;

        if( layout === 'left') {
            content = (
                <View style={[styles.container, style]}>
                    {icon_n}
                    {title_n}
                </View>
            );
        }
        else{
            content = (
                <View  style={[styles.container, style]}>
                    {icon_n}
                    {title_n}
                </View>
            );
        }
        return content;
    },
});

let styles = StyleSheet.create({
    container:{
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
        marginTop: 14,
        marginLeft: 14,
    },
    icon:{
        width: 10,
        height: 17,
    },
    title: {
        marginLeft: 4,
        fontSize: 17,
        color: '#ffffff',
    },
});

module.exports = OrionNavigatorImageButton;
