// react-native-orion/lib/component/navigator/navigatortitle.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    StyleSheet,
    Text,
} from 'react-native';

import OrionComponent from '../component';
import OrionText from '../text/text';

const OrionNavigatorTitle = React.createClass({
    propTypes: {
        ...OrionComponent.propTypes,
        style: Text.propTypes.style,
        title: React.PropTypes.string,
    },
    getDefaultProps: function() {
        return {
            ...OrionText.defaultProps,
        };
    },
    render: function () {
        var {title, style, ...props} = this.props;
        return (
            <OrionText style={[styles.title, style]} l_text={title}/>
        );
    },
});

let styles = StyleSheet.create({
    title: {
        textAlign: 'center',
        fontSize: 17,
        color: '#ffffff',
        marginTop: 14,
    },
});

module.exports = OrionNavigatorTitle;
