// react-native-orion/lib/component/navigator/navigatorbutton.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    View,
} from 'react-native';

import OrionComponent from '../component';
import OrionNavigatorIconFontButton from './navigatoriconfontbtn';
import OrionNavigatorImageButton from './navigatorimagebtn';

const OrionNavigatorButton = React.createClass({
    statics: {
        ImageButton: OrionNavigatorImageButton,
        IconFontButoon: OrionNavigatorIconFontButton,
    },
    
    propTypes: {
        ...OrionComponent.propTypes,
    },
    getDefaultProps: function() {
        return {
            ...OrionComponent.defaultProps,
        };
    },
    render: function () {
        return null;
    },
});

module.exports = OrionNavigatorButton;
