// react-native-orion/lib/component/navigator/titlebar.js
/**
 * Created by unique on 16/7/26.
 */

import React, { Component } from 'react';
import {
    Platform,
    StyleSheet,
    Text,
    ColorPropType,
} from 'react-native';

import {
    Aspect,
    OrionView,
    OrionText,
    OrionComponent,
} from 'react-native-lark';

const OrionTitleBar = React.createClass({
    propTypes: {
        ...OrionComponent.propTypes,
        l_barTintColor: ColorPropType,
        l_tintColor: ColorPropType,
        l_title: React.PropTypes.string,
        l_height: React.PropTypes.number,
    },
    getDefaultProps: function() {
        return {
            ...OrionComponent.defaultProps,
            l_height: 64,
            l_tintColor: '#FFFFFF',
            l_barTintColor: '#000000',
        };
    },
    render:function (){
        var statusBar = 20;
        var height = Platform.OS === 'ios' ? this.props.l_height : this.props.l_height - statusBar;
        var top = Platform.OS === 'ios' ? statusBar : 0;
        return (
            <OrionView style={[{height: height, backgroundColor: this.props.l_barTintColor}, styles.container]}>
                <OrionText l_text={this.props.l_title} style={[{color: this.props.l_tintColor}, styles.title]}/>
            </OrionView>
        );
    }
});

const styles = StyleSheet.create({
    container:{
        flexDirection: 'row',
        alignItems: 'center',
        paddingTop: Platform.OS === 'ios' ? 20 : 0,
    },
    title:{
        textAlign: 'center',
        flex:1,
        fontSize: 17,
    },
});

module.exports = OrionTitleBar;
