// react-native-orion/lib/component/activityindicator/activityindicator.js
/**
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    ActivityIndicator,
    ColorPropType,
} from 'react-native';

import OrionView from '../view/view';

const OrionActivityIndicator = React.createClass({
    propTypes: {
        ...OrionView.propTypes,
        animating: PropTypes.bool,
        color: ColorPropType,
        size: PropTypes.oneOf([
            'small',
            'large',
        ]),
    },
    getDefaultProps: function() {
        return {
            ...OrionView.defaultProps,
            animating: true,
            color: 'gray',
            size: 'small',
        };
    },
    render: function () {
        return (
            <ActivityIndicator {...this.props}/>
        );
    },
});

module.exports = OrionActivityIndicator;
