// react-native-orion/lib/component/image/webimage.ios.js
/**
 * Copyright (c) 2016-present, Unique, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @providesModule OrionWebImage
 * @flow
 */

'use strict';

import React from 'react';
import {
    StyleSheet,
    StyleSheetPropType,
    Image,
    View,
    ViewAttributes,
    requireNativeComponent,
} from 'react-native';

import OrionComponent from '../component';
import TimerMixin from '../../timermixin';

const OrionWebImage = React.createClass({
  mixins: [TimerMixin],

  propTypes: {
    ...OrionComponent.propTypes,
    style: Image.propTypes.style,
    url: React.PropTypes.string,
    fillMode: React.PropTypes.oneOf(['fill','aspectfit','aspectfill']),
  },

  viewConfig: {...View.viewConfig},

  getDefaultProps: function() {
    return {
      ...OrionComponent.defaultProps,
      fillMode: 'aspectfill',
    };
  },

  render: function () {
    let {url, fillMode, style, ...props} = this.props;

    return (
        <RCTSDWebImageView
            {...props}
            style={[styles.base, style]}
            url={url}
            fillMode={fillMode}
        />
    );
  },
});

var styles = StyleSheet.create({
  base: {
    overflow: 'hidden',
  },
});

var RCTSDWebImageView = requireNativeComponent('RCTSDWebImageView', OrionWebImage);

module.exports = OrionWebImage;
