// react-native-orion/lib/component/image/webimage.android.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    Image,
} from 'react-native';

import OrionComponent from '../component';

const OrionWebImage = React.createClass({
  propTypes: {
    ...OrionComponent.propTypes,
    style: Image.propTypes.style,
    l_uri: React.PropTypes.string,
    l_mode: React.PropTypes.oneOf(['fill','aspectfit','aspectfill']),
  },
  getDefaultProps: function() {
    return {
      ...OrionComponent.defaultProps,
      l_mode: 'aspectfill',
    };
  },
  render: function () {
    let {l_uri, l_mode, ...props} = this.props;
    var src={};
    if( l_uri  && l_uri != '' ){
        src.uri = l_uri;
    }
    return (
        <Image {...props} source={src}/>
    );
  },
});

module.exports = OrionWebImage;
