// react-native-orion/lib/component/tabbar/tabbaritem.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    StyleSheet,
    Platform,
    Text,
    View,
    Image,
    TabBarIOS,
    ColorPropType,
} from 'react-native';

import OrionComponent from '../component';

const OrionTabBarItem = React.createClass({
    propTypes: {
        ...OrionComponent.propTypes,
        ...View.propTypes,
        badge: React.PropTypes.oneOfType([
            React.PropTypes.string,
            React.PropTypes.number,
        ]),
        badgeColor: ColorPropType,
        icon: Image.propTypes.source,
        selectedIcon: Image.propTypes.source,
        onPress: React.PropTypes.func,
        selected: React.PropTypes.bool,
        style: View.propTypes.style,
        title: React.PropTypes.string,
    },
    getDefaultProps: function() {
        return {
            ...OrionComponent.defaultProps,
            ...View.defaultProps,
            selectedTab: '',
            badgeColor: '#FF0000'
        };
    },

    renderAndroid: function () {
        let child = React.Children.only(this.props.children);
        return React.cloneElement(child, {
            style:[child.props.style, this.props.style],
        });

    },

    renderIOS: function () {
        return (
            <TabBarIOS.Item
                {...this.props}>
                {this.props.children}
            </TabBarIOS.Item>
        );
    },

    render: function () {
        return  Platform.OS === 'android' ? this.renderAndroid() : this.renderIOS();
        //return this.renderAndroid();
    },
});

module.exports = OrionTabBarItem;
