// react-native-orion/lib/component/tabbar/tabitem.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';
import React from 'react';

import {
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import Aspect from '../../aspect';
import Layout from './layout';

const OrionTabItem  = React.createClass({
  propTypes: {
    title: React.PropTypes.string,
    titleStyle: Text.propTypes.style,
    badge: React.PropTypes.element,
    onPress: React.PropTypes.func,
    hidesTabTouch: React.PropTypes.bool,
    allowFontScaling: React.PropTypes.bool,
    style: View.propTypes.style,
  },
  componentDidMount:function(){
    this._handlePress = this._handlePress.bind(this);
  },

  render:function() {
    let { title, badge } = this.props;
    let icon = null;
    if (React.Children.count(this.props.children) > 0) {
      icon = React.Children.only(this.props.children);
    }

    if (title && title != "" ) {
      title =
        <Text
          numberOfLines={1}
          allowFontScaling={!!this.props.allowFontScaling}
          style={styles.title}>
          {title}
        </Text>;
    }
    else{
      title = null;
    }

    if (badge) {
      badge = React.cloneElement(badge, {
        style: [styles.badge, badge.props.style],
      });
    }

    let tabStyle = [
      styles.container,
      title ? null : styles.untitledContainer,
      this.props.style,
    ];
    return (
      <TouchableOpacity
        activeOpacity={this.props.hidesTabTouch ? 1.0 : 0.8}
        onPress={this._handlePress}
        style={tabStyle}>
        <View>
          {icon}
          {badge}
        </View>
        {title}
      </TouchableOpacity>
    );
  },

  _handlePress(event) {
    if (this.props.onPress) {
      this.props.onPress(event);
    }
  },
});

let styles = StyleSheet.create({
  badge: {
    position: 'absolute',
    top: -6,
    right: -12,
  },
  container: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  untitledContainer: {
    paddingBottom: 17,
  },
  title: {
    color: '#929292',
    fontSize: 10,
    textAlign: 'center',
    alignSelf: 'stretch',
    marginTop: 6,
    marginBottom: 2 + Aspect.onePixel,
  },
});

module.exports = OrionTabItem;
