// react-native-orion/lib/component/view/frameview.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    Platform,
    StyleSheet,
    View,
    ColorPropType,
} from 'react-native';

import Aspect from '../../aspect';
import OrionView from './../component';

const styles = StyleSheet.create({
    container:{
    }
});

const OrionFrameView = React.createClass({
    propTypes: {
        ...OrionView.propTypes,
        l_top: React.PropTypes.number,
        l_left:React.PropTypes.number,
        l_right: React.PropTypes.number,
        l_bottom:React.PropTypes.number,
        l_color: ColorPropType,
    },

    getDefaultProps: function() {
        return {
            ...OrionView.defaultProps,
            l_top: -1,
            l_left: -1,
            l_right: -1,
            l_bottom: -1,
            l_color: "#000000",
        };
    },

    render: function () {
        let {l_top, l_left, l_right, l_bottom, style, l_color, ...props} = this.props;
        var newstyle = {...style};
        var pixelWidth = Platform.OS === 'ios' ? Aspect.onePixel : 1;

        if( l_top >= 0) {
            newstyle.borderTopWidth = l_top == 0 ? pixelWidth : l_top;
            newstyle.borderTopColor = l_color;
        }

        if( l_bottom >= 0) {
            newstyle.borderBottomWidth = l_bottom == 0 ? pixelWidth : l_bottom;
            newstyle.borderBottomColor = l_color;
        }

        if( l_left >= 0) {
            newstyle.borderLeftWidth = l_left == 0 ? pixelWidth : l_left;
            newstyle.borderLeftColor = l_color;
        }

        if( l_right >= 0) {
            newstyle.borderRightWidth = l_right == 0 ? pixelWidth : l_right;
            newstyle.borderRightColor = l_color;
        }

        return (
            <View {...props} style={[styles.container, newstyle]}/>
        );
    },
});

module.exports = OrionFrameView;
