// react-native-orion/lib/component/view/lineview.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    Platform,
    StyleSheet,
    View,
    ColorPropType,
} from 'react-native';

import Aspect from '../../aspect';
import OrionComponent from './../component';

const styles = StyleSheet.create({
   container:{
       position: 'absolute',
   }
});

const OrionLineView = React.createClass({
    propTypes: {
        ...OrionComponent.propTypes,
        l_begin: React.PropTypes.number,
        l_end: React.PropTypes.number,
        l_direction: React.PropTypes.oneOf(['row', 'column']),
        l_color:ColorPropType,
        l_width:React.PropTypes.number,
        l_pos:React.PropTypes.oneOf(['start', 'end']),
    },
    getDefaultProps: function() {
        return {
            ...OrionComponent.defaultProps,
            l_color: '#000000',
            l_width: Platform.OS === 'ios' ? Aspect.onePixel : 1,
            l_pos:'end',
            l_direction:'column',
            l_end:0,
            l_begin:0,
        };
    },
    render: function () {
        let {l_direction, l_color, l_width, l_pos, l_begin, l_end, ...props} = this.props;
        var newstyle = {};

        newstyle.backgroundColor = l_color;

        if( l_direction === 'row'){
            newstyle.width = l_width;
            newstyle.top = l_begin;
            newstyle.bottom = l_end;

            if( l_pos === 'start'){
                newstyle.left = 0;
            }
            else{
                newstyle.right = 0;
            }
        }
        else{
            newstyle.height = l_width;
            newstyle.left = l_begin;
            newstyle.right = l_end;

            if( l_pos === 'start'){
                newstyle.top = 0;
            }
            else{
                newstyle.bottom = 0;
            }
        }

        return (
            <View {...props} style={[styles.container, newstyle]}/>
        );
    },
});

module.exports = OrionLineView;
