// react-native-orion/lib/component/view/spacer.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    StyleSheet,
    View,
} from 'react-native';

import Aspect from '../../aspect';
import OrionComponent from './../component';

const styles = StyleSheet.create({
    container:{
        backgroundColor: 'transparent',
        alignSelf: 'stretch',
    }
});

const OrionSpacer = React.createClass({
    propTypes: {
        ...OrionComponent.propTypes,
        l_direction: React.PropTypes.oneOf(['row', 'column']),
        l_space:React.PropTypes.number,
        l_flex:React.PropTypes.number,
    },
    getDefaultProps: function() {
        return {
            ...OrionView.defaultProps,
            l_flex: 0,
            l_width: 0,
            l_direction:'column',
        };
    },
    render: function () {
        let {l_direction,  l_space, l_flex,  ...props} = this.props;

        var newstyle = {};

        if( l_flex > 0 ){
            newstyle.flex = l_flex;
        }
        else{
            if( l_direction === 'row'){
                newstyle.width = l_space;
            }
            else{
                newstyle.height = l_space;
            }
        }

        return (
            <View {...props} style={[styles.container, newstyle]}/>
        );
    },
});

module.exports = OrionSpacer;
