// react-native-orion/lib/component/view/touchviewPropType.js
/*
 *  Copyright (c) 2016-present, Unique, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

import React from 'react';
import {
    View,
} from 'react-native';

import OrionComponent from  '../component';

const OrionViewPropType = React.PropTypes.shape({
    ...OrionComponent.propTypes,
    ...View.propTypes,

    l_touch: React.PropTypes.shape({
        disabled: React.PropTypes.bool,
        onPress: React.PropTypes.func,
        onPressIn: React.PropTypes.func,
        onPressOut: React.PropTypes.func,
        onLayout: React.PropTypes.func,
        onLongPress: React.PropTypes.func,
        delayPressIn: React.PropTypes.number,
        delayPressOut: React.PropTypes.number,
        delayLongPress: React.PropTypes.number,
        activeOpacity: React.PropTypes.number,
    }),
    underlayColor: ColorPropType,
    l_type: React.PropTypes.oneOf(['highlight','opacity']),
});

module.exports = OrionViewPropType;
